### ISSUE TITLE
wiki: CodeReviewComments#NamedResultParameters

### ISSUE THREAD

**Issue by jkopczyn** [golang/go#32671]
> On the other hand, [...] if the meaning of a result isn't clear from context, adding names may be useful in some contexts. 

In general, this will almost always be true. Humans are extremely bad at assessing how obvious an inference is if they already know the fact to be inferred. This is a special case of the [illusion of transparency](https://en.wikipedia.org/wiki/Illusion_of_transparency). In general, the meaning of a result isn't clear from context even if it feels to the author like it's extremely clear from context.

Because of this, the advice given in this section is harmful in most cases.

**Comment by mvdan** [comment]
If this were true, then most funcs in godoc pages would be confusing, as one wouldn't be able to tell what each result parameter is for. In practice, I don't think that's true.

If you argue that people tend to take the easy route and name result parameters slightly less often than they should, perhaps I'd agree there. But that's entirely different from the wiki advice being harmful.

**Comment by jkopczyn** [comment]
In my experience it is absolutely true in practice. Godoc pages are mostly
devoid of usable information unless they have details in docstrings.

**Comment by agnivade** [comment]
@andybons @dmitshur

**Comment by andybons** [comment]
Thanks for the suggestion.

Naming the result parameter doesn’t make the following function more clear:

`func TrimSpace(s string) string`

The [accompanying godoc](https://golang.org/pkg/strings/#TrimSpace) is sufficient enough to further understand what the function is doing.

Sorry, but the advice in the wiki page will remain as is. We don’t want to encourage people to use named return values all the time.

**Comment by jkopczyn** [comment]
You didn't actually engage with my point. Sure, in tiny one-result functions there are some where named returns don't help. That doesn't mean the majority or even plurality of functions are places where named returns are unhelpful.

**Comment by mvdan** [comment]
@jkopczyn you're saying that most functions require named result parameters to not be confusing, and you were given a simple counter-example - I think that's a fair reply. After all, you haven't given one example of a real function that's confusing because it lacks names.

If you want to prove your point, consider backing it up with data. For example, take a few std packages, count the ratio of functions that use named parameters, and give examples of functions with unnamed parameters that are confusing.

**Comment by jkopczyn** [comment]
No, that's not what I said. Most functions are more confusing if they leave
the returns unnamed; this is supported by the rational prior from
psychology.
Good docstrings can compensate for this, and the standard library largely
has good docstrings. Its functions would still be easier to understand if
they defaulted to naming their return values.

**Comment by mvdan** [comment]
Sorry, but this is still hard to follow without any data or examples. You also mention that good godoc strings can help here, but your original suggestion seems to leave them out of the equation.

Note that I don't have a strong opinion here, other than not wanting to encourage general use of named result parameters.
